src/ctia/schemas/graphql/helpers.clj

```diff
@@ -29,9 +29,27 @@
             GraphQLUnionType
             TypeResolver]))
 
+;; type-registry is an Atom<{String, IDeref<graphql.*>}>
+(defn get-or-update-type-registry
+  "If name exists in registry, return existing mapping. Otherwise conj
+  {name (delay (f))} to registry, and return @(delay (f))."
+  [type-registry name f]
+  {:post [%]}
+  (or ;; fast-path for readers
+      (some-> (get @type-registry name) deref)
+      ;; generate a new graphql value, or coordinate with another thread doing the same
+      (let [f (bound-fn* f) ;; may be run on another thread
+            {result-delay name} (swap! type-registry
+                                       (fn [{existing-delay name :as oldtr}]
+                                         (cond-> oldtr
+                                           (not existing-delay)
+                                           (assoc name (delay (f))))))]
+        @result-delay)))
```

Comment:
This implementation assumes full parallelism, the possibility that `f` may use thread bindings, and an invariant that `f` can only be called once.